### ISSUE TITLE
the deploy prototxt is right？

### ISSUE THREAD

**Issue by wolfworld6** [NetEase-GameAI/ResNeSt-caffe#1]
hello，are you sure about the results ，was tested with your release prototxt？
F0605 15:50:30.684473 40576 scale_layer.cpp:92] Check failed: bottom[0]->shape(axis_ + i) == scale->shape(i) (512 vs. 3584) dimension mismatch between bottom[0]->shape(1) and scale->shape(1)

**Comment by NetEase-GameAI** [comment]
Hi, we just downloaded these 4 models and checked carefully. For both `$CAFFEROOT/build/tools/caffe` and `pycaffe.forward()`(with prototxt and caffemodel), everything is all right.
Did you use official bvlc-caffe or a customized version?

**Comment by wolfworld6** [comment]
> Hi, we just downloaded these 4 models and checked carefully. For both `$CAFFEROOT/build/tools/caffe` and `pycaffe.forward()`(with prototxt and caffemodel), everything is all right.
> Did you use official bvlc-caffe or a customized version?

yeah，I used the official bvlc-caffe，in your deploy prototxt，the "reshape422" blob is [10,4096], then the next layer, "slice423" blob0:[10,512],blob1[10,3584],why?

**Comment by NetEase-GameAI** [comment]
It seems you got wired results on reshape422. Since pool412 outputs NxCx1x1 and no pad in conv413 and conv416, the expected shapes should be: conv416(Nx1024x1x1), reshape418(Nx1x2x512), perm419(Nx2x1x512), reshape422(Nx1024).

Here are our results when setting batch_size=10. 

I0608 11:14:23.897440  3550 layer_factory.hpp:77] Creating layer reshape418
I0608 11:14:23.897445  3550 net.cpp:86] Creating Layer reshape418
I0608 11:14:23.897449  3550 net.cpp:408] reshape418 <- conv416
I0608 11:14:23.897452  3550 net.cpp:382] reshape418 -> reshape418
I0608 11:14:23.897459  3550 net.cpp:124] Setting up reshape418
I0608 11:14:23.897475  3550 net.cpp:131] Top shape: 10 1 2 512 (10240)
I0608 11:14:23.897487  3550 net.cpp:139] Memory required for data: 3640558080
I0608 11:14:23.897501  3550 layer_factory.hpp:77] Creating layer perm419
I0608 11:14:23.897519  3550 net.cpp:86] Creating Layer perm419
I0608 11:14:23.897533  3550 net.cpp:408] perm419 <- reshape418
I0608 11:14:23.897549  3550 net.cpp:382] perm419 -> perm419
I0608 11:14:23.897570  3550 net.cpp:124] Setting up perm419
I0608 11:14:23.897585  3550 net.cpp:131] Top shape: 10 2 1 512 (10240)
I0608 11:14:23.897598  3550 net.cpp:139] Memory required for data: 3640599040
I0608 11:14:23.897612  3550 layer_factory.hpp:77] Creating layer softmax420
I0608 11:14:23.897629  3550 net.cpp:86] Creating Layer softmax420
I0608 11:14:23.897641  3550 net.cpp:408] softmax420 <- perm419
I0608 11:14:23.897660  3550 net.cpp:382] softmax420 -> softmax420
I0608 11:14:23.897677  3550 net.cpp:124] Setting up softmax420
I0608 11:14:23.897692  3550 net.cpp:131] Top shape: 10 2 1 512 (10240)
I0608 11:14:23.897706  3550 net.cpp:139] Memory required for data: 3640640000
I0608 11:14:23.897727  3550 layer_factory.hpp:77] Creating layer reshape422
I0608 11:14:23.897742  3550 net.cpp:86] Creating Layer reshape422
I0608 11:14:23.897758  3550 net.cpp:408] reshape422 <- softmax420
I0608 11:14:23.897773  3550 net.cpp:382] reshape422 -> reshape422
I0608 11:14:23.897791  3550 net.cpp:124] Setting up reshape422
I0608 11:14:23.897806  3550 net.cpp:131] Top shape: 10 1024 (10240)
I0608 11:14:23.897820  3550 net.cpp:139] Memory required for data: 3640680960
I0608 11:14:23.897835  3550 layer_factory.hpp:77] Creating layer slice423
I0608 11:14:23.897857  3550 net.cpp:86] Creating Layer slice423
I0608 11:14:23.897871  3550 net.cpp:408] slice423 <- reshape422
I0608 11:14:23.897886  3550 net.cpp:382] slice423 -> slice423-0
I0608 11:14:23.897903  3550 net.cpp:382] slice423 -> slice423-1
I0608 11:14:23.897920  3550 net.cpp:124] Setting up slice423
I0608 11:14:23.897935  3550 net.cpp:131] Top shape: 10 512 (5120)
I0608 11:14:23.897949  3550 net.cpp:131] Top shape: 10 512 (5120)
I0608 11:14:23.897963  3550 net.cpp:139] Memory required for data: 3640721920

**Comment by wolfworld6** [comment]
> It seems you got wired results on reshape422. Since pool412 outputs NxCx1x1 and no pad in conv413 and conv416, the expected shapes should be: conv416(Nx1024x1x1), reshape418(Nx1x2x512), perm419(Nx2x1x512), reshape422(Nx1024).
> 
> Here are our results when setting batch_size=10.
> 
> I0608 11:14:23.897440 3550 layer_factory.hpp:77] Creating layer reshape418
> I0608 11:14:23.897445 3550 net.cpp:86] Creating Layer reshape418
> I0608 11:14:23.897449 3550 net.cpp:408] reshape418 <- conv416
> I0608 11:14:23.897452 3550 net.cpp:382] reshape418 -> reshape418
> I0608 11:14:23.897459 3550 net.cpp:124] Setting up reshape418
> I0608 11:14:23.897475 3550 net.cpp:131] Top shape: 10 1 2 512 (10240)
> I0608 11:14:23.897487 3550 net.cpp:139] Memory required for data: 3640558080
> I0608 11:14:23.897501 3550 layer_factory.hpp:77] Creating layer perm419
> I0608 11:14:23.897519 3550 net.cpp:86] Creating Layer perm419
> I0608 11:14:23.897533 3550 net.cpp:408] perm419 <- reshape418
> I0608 11:14:23.897549 3550 net.cpp:382] perm419 -> perm419
> I0608 11:14:23.897570 3550 net.cpp:124] Setting up perm419
> I0608 11:14:23.897585 3550 net.cpp:131] Top shape: 10 2 1 512 (10240)
> I0608 11:14:23.897598 3550 net.cpp:139] Memory required for data: 3640599040
> I0608 11:14:23.897612 3550 layer_factory.hpp:77] Creating layer softmax420
> I0608 11:14:23.897629 3550 net.cpp:86] Creating Layer softmax420
> I0608 11:14:23.897641 3550 net.cpp:408] softmax420 <- perm419
> I0608 11:14:23.897660 3550 net.cpp:382] softmax420 -> softmax420
> I0608 11:14:23.897677 3550 net.cpp:124] Setting up softmax420
> I0608 11:14:23.897692 3550 net.cpp:131] Top shape: 10 2 1 512 (10240)
> I0608 11:14:23.897706 3550 net.cpp:139] Memory required for data: 3640640000
> I0608 11:14:23.897727 3550 layer_factory.hpp:77] Creating layer reshape422
> I0608 11:14:23.897742 3550 net.cpp:86] Creating Layer reshape422
> I0608 11:14:23.897758 3550 net.cpp:408] reshape422 <- softmax420
> I0608 11:14:23.897773 3550 net.cpp:382] reshape422 -> reshape422
> I0608 11:14:23.897791 3550 net.cpp:124] Setting up reshape422
> I0608 11:14:23.897806 3550 net.cpp:131] Top shape: 10 1024 (10240)
> I0608 11:14:23.897820 3550 net.cpp:139] Memory required for data: 3640680960
> I0608 11:14:23.897835 3550 layer_factory.hpp:77] Creating layer slice423
> I0608 11:14:23.897857 3550 net.cpp:86] Creating Layer slice423
> I0608 11:14:23.897871 3550 net.cpp:408] slice423 <- reshape422
> I0608 11:14:23.897886 3550 net.cpp:382] slice423 -> slice423-0
> I0608 11:14:23.897903 3550 net.cpp:382] slice423 -> slice423-1
> I0608 11:14:23.897920 3550 net.cpp:124] Setting up slice423
> I0608 11:14:23.897935 3550 net.cpp:131] Top shape: 10 512 (5120)
> I0608 11:14:23.897949 3550 net.cpp:131] Top shape: 10 512 (5120)
> I0608 11:14:23.897963 3550 net.cpp:139] Memory required for data: 3640721920

ok, on conv413-bn414-scale414, are you sure about both the sacle and the conv "bias_term: true"?
![image](https://user-images.githubusercontent.com/18051979/83990491-73ad5680-a97c-11ea-85aa-b962a9df7fa2.png)

**Comment by NetEase-GameAI** [comment]
Yes, both conv413 and scale414 need `bias_term`. According to BN calculation rules, Scale layer need `bias_term`. For Convolution+BN layer, it's ok to drop `bias_term` in Convolution layer.
However, since the `fc1` layer in https://github.com/zhanghang1989/ResNeSt/blob/master/resnest/torch/splat.py#L39 use default bias setting(bias=True), we just keep it in conv413.

**Comment by wolfworld6** [comment]
> Yes, both conv413 and scale414 need `bias_term`. According to BN calculation rules, Scale layer need `bias_term`. For Convolution+BN layer, it's ok to drop `bias_term` in Convolution layer.
> However, since the `fc1` layer in https://github.com/zhanghang1989/ResNeSt/blob/master/resnest/torch/splat.py#line39 use default bias setting(bias=True), we just keep it in conv413.

Thank you so much for your reply,I will try !
